### PR TITLE
added nearest neighbor algorithm - machine-learning

### PR DESCRIPTION
**Pull request by christianbender**
* Two functions ```nearest_neighbor(x, tSet)``` and a (eulidean) distance function ```distance(x,y)```.   
* I used the ```numpy``` library for calculating the absolute value of a vector.  
* Two trainings sets for the logical AND-function and for color-analysing (dark/light color).  

Have someone a tip for finding the ***k-nearest*** neighbors? I wrote a simple algorithm.  

- [X] **If creating a new file :**
  - [X] added links to it in the README files ?
  - [X] included tests with it ?
  - [ ] added description (overview of algorithm, time and space compleixty, and possible edge case) in docstrings ?

- [ ] **if done some changes :**
  - [ ] wrote short description in the PR explaining what the changes do ?
  - [ ] Fixes #[issue number] if related to any issue

- [] **other**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -350,6 +350,8 @@ If you want to uninstall algorithms, it is as simple as:
         - [simplify_path](algorithms/unix/path/simplify_path.py)
 - [union-find](algorithms/union-find)
     - [count_islands](algorithms/union-find/count_islands.py)
+- [machine-learning](algorithms/machine-learning)
```
Comment (keon) [inline README.md]: lets change the folder name to ml for simplicity.
As a package, we need to think how people are going to use it.
typing `algorithms.ml` is better than typing `algorithms.machine-learning` every time.

File: algorithms/machine-learning/nearest_neighbor.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import numpy
+
+# train set for the AND-function
+trainSetAND = {(0,0) : 0, (0,1) :0, (1,0) : 0, (1,1) : 1} 
+
+# train set for light or dark colors
+trainSetLight = {(11, 98, 237) : 'L', (3, 39, 96) : 'D', (242, 226, 12) : 'L', (99, 93, 4) : 'D',
```
Comment (keon) [inline algorithms/machine-learning/nearest_neighbor.py]: Let's separate the test to the test folder.

File: algorithms/machine-learning/nearest_neighbor.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    for key in tSet:
+        d = distance(x, key)
+        if d < min_d:
+            min_d = d
+            current_key = key
+    return tSet[current_key]
+
+# Some test cases
+
+# print(nearest_neighbor((1,1), trainSetAND)) # => 1
```
Comment (keon) [inline algorithms/machine-learning/nearest_neighbor.py]: same here.

File: algorithms/machine-learning/nearest_neighbor.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    Arguments:
+        x {[tupel]} -- [vector]
+        tSet {[dict]} -- [training set]
+    
+    Returns:
+        [type] -- [result of the AND-function]
+    """
+    assert isinstance(x, tuple) and isinstance(tSet, dict)
+    current_key = ()
+    MAX = 32768 # max value
```
Comment (keon) [inline algorithms/machine-learning/nearest_neighbor.py]: is there a reason for setting this to 32768?

Comment (christianbender) [reply algorithms/machine-learning/nearest_neighbor.py]: @keon I want a high max value.

File: README.md (hunk 1 above)
Comment (christianbender) [reply README.md]: @keon Good point! I rename the directory

**Comment by christianbender** [comment]
@keon Done.

File: algorithms/ml/nearest_neighbor.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    calculates the (eulidean) distance between vector x and y.
+    
+    We use the numpy libriary
+
+    Arguments:
+        x {[tuple]} -- [vector]
+        y {[tuple]} -- [vector]
+    """
+    assert len(x) == len(y), "The vector must have same length"
+    import math
```
Comment (goswami-rahul) [inline algorithms/ml/nearest_neighbor.py]: why import math?

File: algorithms/ml/nearest_neighbor.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import numpy
```
Comment (goswami-rahul) [inline algorithms/ml/nearest_neighbor.py]: We are using a 3rd party module here. `numpy` will become this repo's requirement/dependency, which is not desirable. I think we can implement these without using `numpy`, unless it is absolutely necessary (and I think that won't be the case), as these are meant to be _minimal_ algorithms in Python.  
By the way, great idea to implement ML algorithms here. I will also add some when I get time:)

File: algorithms/machine-learning/nearest_neighbor.py (hunk 4 above)
Comment (christianbender) [reply algorithms/machine-learning/nearest_neighbor.py]: I changed it. Thanks

**Comment by christianbender** [comment]
@keon @goswami-rahul @danghai  @SaadBenn 
Usage of the module ```numpy```? 
This module it wide use in machine learning in python. 
I can implement the the parts (above) by myself, too.

File: algorithms/ml/nearest_neighbor.py (hunk 5 above)
Comment (christianbender) [reply algorithms/ml/nearest_neighbor.py]: @goswami-rahul for the function ```sqrt(...)``` that I used before ```numpy```.

File: algorithms/ml/nearest_neighbor.py (hunk 6 above)
Comment (christianbender) [reply algorithms/ml/nearest_neighbor.py]: Here free online course on edx : https://www.edx.org/course/machine-learning-fundamentals-uc-san-diegox-dse220x

**Comment by danghai** [comment]
@christianbender 
It is nice to implement both :). The one is how to use the API from library, and another one is how to write them.

**Comment by goswami-rahul** [comment]
@christianbender @keon @danghai 
IMHO, a user who wants to install `algorithms` for some specific algorithms, say arrays or dp or strings, etc. He might not want to install `numpy` or `pandas` or any other high level libraries for his requirements.  
Moreover, I think this repo's objective is to provide __minimal__ implementations of the algorithms. These algorithms are not very difficult to implement using built-in Python libraries, even if it would be less efficient. I guess we should be going with Simplicity over Efficiency, but it is just a matter of opinion :) The collections of super efficient and optimized algorithms are already available and are constantly being updated, but ours could be the simplest, understandable and minimal collection in Python from scratch!  
Just my two cents. Cheers :)
Please share your thoughts on this.

**Comment by christianbender** [comment]
@goswami-rahul 

> IMHO, a user who wants to install algorithms for some specific algorithms, say arrays or dp or strings, etc. He might not want to install numpy or pandas or any other high level libraries for his requirements.

Good point! :+1: 

@goswami-rahul @danghai @keon 

I will change my code again.

**Comment by christianbender** [comment]
@danghai @goswami-rahul @keon 
Done. I removed ```numpy```

**Comment by danghai** [comment]
@christianbender Could you recheck the `travis`? It is fail in python3.4 and 3.6

**Comment by danghai** [comment]
@christianbender click `Details` at Travis test below. It directly goes : https://travis-ci.org/keon/algorithms/builds/409541731?utm_source=github_status&utm_medium=notification

**Comment by danghai** [comment]
@christianbender I fix this issue for you. Look my above commit.

**Comment by christianbender** [comment]
@danghai Thanks a lot.

**Comment by christianbender** [comment]
@danghai I merge it.

**Pull request merged**